servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/TimeoutHttpRequesterFilter.java

```diff
@@ -30,46 +34,109 @@
+     * Creates a new instance which requires only that the response metadata be received before the timeout.
      *
      * @param duration the timeout {@link Duration}
      */
     public TimeoutHttpRequesterFilter(final Duration duration) {
-        this.duration = duration;
-        this.timeoutExecutor = null;
+        this(useDefaultTimeout(duration), false);
     }
 
     /**
-     * Creates a new instance.
+     * Creates a new instance which requires only that the response metadata be received before the timeout.
      *
      * @param duration the timeout {@link Duration}
      * @param timeoutExecutor the {@link Executor} to use for managing the timer notifications
      */
     public TimeoutHttpRequesterFilter(final Duration duration, final Executor timeoutExecutor) {
-        this.duration = duration;
-        this.timeoutExecutor = requireNonNull(timeoutExecutor);
+        this(useDefaultTimeout(duration), false, timeoutExecutor);
+    }
+
+    /**
+     * Creates a new instance.
+     *
+     * @param timeoutForRequest function for extracting timeout from request
+     * @param fullRequestResponse if true then timeout is for full request/response transaction otherwise only the
+     * response metadata must arrive before the timeout.
+     */
+    public TimeoutHttpRequesterFilter(final TimeoutFromRequest timeoutForRequest, final boolean fullRequestResponse) {
+        this.timeoutForRequest = timeoutForRequest;
+        this.fullRequestResponse = fullRequestResponse;
+        this.timeoutExecutor = null;
+    }
+
+    /**
+     * Creates a new instance.
+     *
+     * @param timeoutForRequest function for extracting timeout from request
+     * @param fullRequestResponse if true then timeout is for full request/response transaction otherwise only the
+     * response metadata must arrive before the timeout.
+     * @param timeoutExecutor the {@link Executor} to use for managing the timer notifications
+     */
+    public TimeoutHttpRequesterFilter(final TimeoutFromRequest timeoutForRequest,
+                                      final boolean fullRequestResponse,
+                                      final Executor timeoutExecutor) {
+        this.timeoutForRequest = timeoutForRequest;
+        this.fullRequestResponse = fullRequestResponse;
+        this.timeoutExecutor = timeoutExecutor;
     }
 
     private Single<StreamingHttpResponse> request(final StreamingHttpRequester delegate,
                                                   final HttpExecutionStrategy strategy,
                                                   final StreamingHttpRequest request) {
-        return timeoutExecutor != null ? delegate.request(strategy, request).timeout(duration, timeoutExecutor) :
-                delegate.request(strategy, request).timeout(duration);
+
+        Single<StreamingHttpResponse> response = delegate.request(strategy, request);
+
+        Duration timeout = timeoutForRequest.apply(request);
```

Comment:
the server has this function invoked inside the `defer` which means it will be re-invoked on each re-subscribe/retry, but here it is outside the scope of `defer`. can we make them consistent?